ui/lib/core/addon/decorators/confirm-leave.js

```diff
@@ -0,0 +1,46 @@
+import { action } from '@ember/object';
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+/**
+ * Confirm that the user wants to discard unsaved changes before leaving the page.
+ * This decorator hooks into the willTransition action. If you override setupController,
+ * be sure to set 'model' on the controller to store data or this won't work.
+ */
+export function withConfirmLeave() {
+  return function decorator(SuperClass) {
+    if (!Object.prototype.isPrototypeOf.call(Route, SuperClass)) {
+      // eslint-disable-next-line
+      console.error(
+        'withConfirmLeave decorator must be used on instance of ember Route class. Decorator not applied to returned class'
+      );
+      return SuperClass;
+    }
+    return class ConfirmLeave extends SuperClass {
+      @service store;
+
+      @action
+      willTransition(transition) {
```

Comment:
I might have had something else going on, but when I tried that the whole thing broke. I'll give it another shot now that it's stable

Reply:
Alright, I updated it to call the super's willTransition within a try/catch because any checks I did to see if it was a callable function resulted in the method not being called. Without the try/catch we get an error that the method doesn't exist

Reply:
Hmmm. I had a look at the docs and they have `willTransition` listed as an event so I guess it isn't actually a method on the `Route` class. We could probably remove the `super` and be safe but I don't think it hurts to have it in the try catch either. Thanks!